### PR TITLE
[Feature] qualify/disqualify/revert a candidate mutations

### LINKED ISSUE #9190: ✨ New graphql mutations for qualifying and disqualifying an application
## ✨ Feature

Create new specific mutations for recording (and reverting) the final decision to qualify or disqualify a Pool Candidate. 
Also add a some metadata to the Candidate model, like decision timestamp.

## 🌎 Context

These mutations will help us deprecate `updatePoolCandidateAsAdmin`, and hide the business logic of Pool Candidate state transitions in helpful ways (eg storing timestamps, and validating the transition to/from states). 

See parent issue #9149 for more context.

### New field

Add a new datetime field to the PoolCandidate model which records when the decision was made. Protect it with the `viewStatus` policy method.

**Question:** the name of this field is up for discussion.

```
type PoolCandidate {
  ...
  finalDecisionAt: DateTime @canOnParent(ability: "viewStatus")
}
```

In the migration, the column should default to null, but then it should go through existing rows and set the timestamp to now() if current status is one of:
    - QUALIFIED_AVAILABLE
    - QUALIFIED_UNAVAILABLE
    - QUALIFIED_WITHDREW
    - PLACED_CASUAL
    - PLACED_TERM
    - PLACED_INDETERMINATE
    - EXPIRED
    - SCREENED_OUT_APPLICATION
    - SCREENED_OUT_ASSESSMENT

### Mutations

These mutations should all check the `PoolCandidatePolicy->updateStatus` policy method

- [ ] `disqualifyCandidate({poolCandidateId!, reason!})`
  - [ ] reason is a new enum, with values representing "screened out on application" and "unsuccessful during assessment"
  - [ ] sets status to SCREENED_OUT_APPLICATION or SCREENED_OUT_ASSESSMENT according to reason
  - [ ] set new finalDecisionAt field with current timestamp
  - [ ] validates that application is currently under assessment - throw an error otherwise. For now, that means status is one of:
    - NEW_APPLICATION
    - APPLICATION_REVIEW
    - SCREENED_IN
    - UNDER_ASSESSMENT

- [ ] `qualifyCandidate({poolCandidateId!, expiryDate!})`
  - [ ] sets status to QUALIFIED_AVAILABLE
  - [ ] sets expiryDate
  - [ ] sets new finalDecisionAt field with current timestamp
  - [ ] validates that application is currently under assessment - throw an error otherwise. For now, that means status is one of:
    - NEW_APPLICATION
    - APPLICATION_REVIEW
    - SCREENED_IN
    - UNDER_ASSESSMENT

- [ ] `revertFinalDecision({poolCandidateId!})`
    - [ ] Set candidate status to NEW_APPLICATION. **Question:** or UNDER_ASSESSMENT?
    - [ ] Nullify expiry date
    - [ ] Nullify finalDecisionAt
    - [ ] Throws an error if candidate is still in assessment, or removed, or placed. Currently, that means status MUST be one of:
      - SCREENED_OUT_APPLICATION
      - SCREENED_OUT_ASSESSMENT
      - QUALIFIED_AVAILABLE
      - EXPIRED

### Use

The Final Decision Modal (#8964) should be updated to use these mutations instead of updatePoolCandidateAsAdmin.

## 🎨 Design File

This covers the qualify, disqualify and revert operations in this state diagram:
Pool Candidate State Transitions

### 🧑‍🎨 Designer

@tristan-orourke 

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] New timestamp field added to database, model and schema
- [ ] The new timestamp has a value for existing rows where it should
- [ ] PoolCandidateFactory sets new timestamp in sensible way

### New mutations:
- [ ] `disqualifyCandidate({poolCandidateId!, reason!})`
  - [ ] reason is a new enum, with values representing "screened out on application" and "unsuccessful during assessment"
  - [ ] sets status to SCREENED_OUT_APPLICATION or SCREENED_OUT_ASSESSMENT according to reason
  - [ ] set new finalDecisionAt field with current timestamp
  - [ ] validates that application is currently under assessment - throw an error otherwise. For now, that means status is one of:
    - NEW_APPLICATION
    - APPLICATION_REVIEW
    - SCREENED_IN
    - UNDER_ASSESSMENT

- [ ] `qualifyCandidate({poolCandidateId!, expiryDate!})`
  - [ ] sets status to QUALIFIED_AVAILABLE
  - [ ] sets expiryDate
  - [ ] sets new finalDecisionAt field with current timestamp
  - [ ] validates that application is currently under assessment - throw an error otherwise. For now, that means status is one of:
    - NEW_APPLICATION
    - APPLICATION_REVIEW
    - SCREENED_IN
    - UNDER_ASSESSMENT

- [ ] `revertFinalDecision({poolCandidateId!})`
    - [ ] Set candidate status to NEW_APPLICATION. **Question:** or UNDER_ASSESSMENT?
    - [ ] Nullify expiry date
    - [ ] Nullify finalDecisionAt
    - [ ] Throws an error if candidate is still in assessment, or removed, or placed. Currently, that means status must be one of:
      - SCREENED_OUT_APPLICATION
      - SCREENED_OUT_ASSESSMENT
      - QUALIFIED_AVAILABLE
      - EXPIRED
- [ ] New migrations all check the `PoolCandidatePolicy->updateStatus` policy method
- [ ] Final Decision Modal uses new mutations instead of updatePoolCandidateAsAdmin
- [ ] phpunit tests for new mutations, which check that status and timestamps are modified correctly, that applicants can't operate on their own applications, and that the mutations fail if coming from the wrong status.

### PR DESCRIPTION
**Pull request by vd1992**
🤖 Resolves #9190

## 👋 Introduction

Adds the three mutations as well as a `final_decision_at` date time field. Includes factory, schema, and testing. 
Also swaps in the new mutations for the final decision dialog. 

## 🧪 Testing

1. Migrate back and forth
2. Seed fresh
3. Test new mutations, see below
4. Use final decision dialog to qualify or disqualify candidates, remember to reset status first in some cases

```
mutation qualifyCandidate {
  qualifyCandidate(
    id: ":ID"
    expiryDate: ":date"
  ) {
    id
    status
    expiryDate
    finalDecisionAt
  }
}
```
```
mutation disqualifyCandidate {
  disqualifyCandidate(
    id: ":id"
    reason: ":reason"  ) {
    id
    status
    finalDecisionAt
  }
}
```
```
mutation revertFinalDecision {
  revertFinalDecision(id: ":id") {
    id
    status
    expiryDate
    finalDecisionAt
  }
}
```

## 📸 Screenshot

image

image

image

### CODE REVIEW

**Review by petertgiles — CHANGES_REQUESTED** [review]
Works great, nice one!  :medal_sports: I noted a few small things to tidy before merging.

File: api/app/GraphQL/Mutations/DisqualifyCandidate.php
Comment (vd1992) [reply api/app/GraphQL/Mutations/DisqualifyCandidate.php]: Sure, done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/10066/commits/3812c6c06350585847c3019850e62f0d0eb9483e and https://github.com/GCTC-NTGC/gc-digital-talent/pull/10066/commits/548fcea3b7ca39f1f2bf6cab0c17f7825ed2f1b8
Decided to adjust the mutations I added recently too for consistency

File: api/tests/Feature/PoolCandidateUpdateTest.php
Comment (vd1992) [reply api/tests/Feature/PoolCandidateUpdateTest.php]: Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/10066/commits/1a967c2a24e9b102b4dce3d8294a941e8250a22c

File: api/database/factories/PoolCandidateFactory.php
Comment (vd1992) [reply api/database/factories/PoolCandidateFactory.php]: Neat, thanks for sharing
Decided to implement it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/10066/commits/2b7c22dad897be9f78958ccde54e8f86f2538502

File: apps/web/src/pages/PoolCandidates/ViewPoolCandidatePage/components/MoreActions/FinalDecisionDialog.tsx
Comment (vd1992) [reply apps/web/src/pages/PoolCandidates/ViewPoolCandidatePage/components/MoreActions/FinalDecisionDialog.tsx:1]: Yeah, I think once Phase II is done operations will be hidden if not applicable

File: api/database/helpers/ApiErrorEnums.php
Comment (vd1992) [reply api/database/helpers/ApiErrorEnums.php:29]: I elected not to do that for a few reasons, let me know what you think
- I wasn't sure copy changes were in scope
- Whether error message copy was decided yet
- There are other error messages too and to group them in one issue to avoid missing any, such as the above errors as well as the ones for revert final decision and the placing pair of errors. So, ensuring all error codes in ErrorEnums have copy in one ticket
- To not block potentially for a while https://github.com/GCTC-NTGC/gc-digital-talent/issues/9197

Comment (petertgiles) [reply api/database/helpers/ApiErrorEnums.php:29]: Yeah, those are all good reasons.  Could you create that issue and make sure it's part of the ROD2 milestone?

**Review by petertgiles — APPROVED** [review]

File: api/database/helpers/ApiErrorEnums.php
Comment (vd1992) [reply api/database/helpers/ApiErrorEnums.php:29]: Issue https://github.com/GCTC-NTGC/gc-digital-talent/issues/10075

**Pull request merged**
